iml-api/src/graphql/stratagem.rs

```diff
@@ -0,0 +1,680 @@
+            if r.0 >= p.0 {
+                return Err(FieldError::new(
+                    "Report duration must be less than Purge duration.",
+                    Value::null(),
+                ));
+            }
+        }
+
+        let uuid = Uuid::new_v4().to_hyphenated().to_string();
+
+        let mut cleanup_tasks = vec![];
+
+        let mut jobs = vec![SendJob {
+            class_name: "ClearOldStratagemDataJob".into(),
+            args: HashMap::<String, serde_json::Value>::new(),
+        }];
+
+        let mut groups = vec!["size_distribution".into(), "user_distribution".into()];
+
+        let fs_id = fs_id_by_name(&context.pg_pool, &fsname).await?;
+
+        if report_duration.is_some() {
+            let task = insert_task(
+                &format!("{}-warn_fids-fids_expiring_soon", uuid),
+                &"created",
+                false,
+                false,
+                &["stratagem.warning".into()],
+                serde_json::json!({
+                    "report_name": format!("expiring_fids-{}-{}.txt", fsname, uuid)
+                }),
+                fs_id,
+                &context.pg_pool,
+            )
+            .await?;
+
+            jobs.push(SendJob {
+                class_name: "CreateTaskJob".into(),
+                args: vec![("task_id".into(), serde_json::json!(task.id))]
+                    .into_iter()
+                    .collect(),
+            });
+
+            cleanup_tasks.push(task);
+
+            groups.push("warn_fids".into());
+        }
+
+        if purge_duration.is_some() {
+            let task = insert_task(
+                &format!("{}-purge_fids-fids_expired", uuid),
+                &"created",
+                false,
+                false,
+                &["stratagem.purge".into()],
+                serde_json::json!({}),
+                fs_id,
+                &context.pg_pool,
+            )
+            .await?;
+
+            jobs.push(SendJob {
+                class_name: "CreateTaskJob".into(),
+                args: vec![("task_id".into(), serde_json::json!(task.id))]
+                    .into_iter()
+                    .collect(),
+            });
+
+            cleanup_tasks.push(task);
+
+            groups.push("purge_fids".into());
+        }
+
+        let job_range: Vec<_> = (0..jobs.len()).collect();
+
+        let xs = get_target_hosts_by_fsname(&fsname, &context.pg_pool).await?;
+
+        for x in xs {
+            let path = match x.dev_path {
+                Some(x) => x,
+                None => continue,
+            };
+
+            let mut cfg = stratagem::StratagemConfig {
+                flist_type: "none".into(),
+                summarize_size: true,
+                device: stratagem::StratagemDevice {
+                    path,
+                    groups: groups.clone(),
+                },
+                groups: vec![
+                    stratagem::StratagemGroup {
```

Comment:
In the old version, generating the size distribution happened every time we did a scan.  Now it's only when report/purge are called.

Reply:
Yes, and I think this is the correct behavior.

The UI for size_distribution and inode users (the only place where this data is visible) is currently tightly coupled to running this particular scan with these particular rules. I don't see a benefit to duplicating these groups in the other mutations as well.

What we should do in a future patch is allow custom stats ingestion and reporting; I think the first step of that is to limit hardcoding groups like these to only the scans where they are truly needed.